src/app/bungie-api/http-client.ts

```diff
@@ -0,0 +1,219 @@
+    serverResponse && (serverResponse as any).error && (serverResponse as any).error_description;
+  if (eMessage) {
+    throw new BungieError(
+      {
+        Message: eMessage,
+        ErrorCode: PlatformErrorCodes.DestinyUnexpectedError,
+        ErrorStatus: eMessage,
+      },
+      request
+    );
+  }
+
+  if (serverResponse.ErrorCode !== PlatformErrorCodes.Success) {
+    throw new BungieError(serverResponse, request);
+  }
+
+  return serverResponse;
+}
+
+//
+// FETCH UTILS
+//
+
+/**
+ * returns a fetch-like that will abort the request after some time
+ *
+ * @param fetchFunction use this function to make the request
+ * @param timeout abort request after this many milliseconds
+ */
+export function createFetchWithTimeout(fetchFunction: typeof fetch, timeout: number): typeof fetch {
+  return async (...[input, init]: Parameters<typeof fetch>) => {
+    const controller = typeof AbortController === 'function' ? new AbortController() : null;
+    const signal = controller?.signal;
+    let timer: NodeJS.Timeout | undefined = undefined;
+
+    if (controller) {
+      timer = setTimeout(() => controller.abort(), timeout);
+      if (typeof input === 'string') {
+        input = new Request(input);
+      }
+      init = { ...init, signal };
+    }
+    try {
+      return await fetchFunction(input, init);
+    } finally {
+      if (timer !== undefined) {
+        clearTimeout(timer);
+      }
+    }
+  };
+}
+
+/**
+ * returns a fetch-like that will run a function if the request is taking a long time,
+ * e.g. generate a "still waiting!" notification
+ *
+ * @param fetchFunction use this function to make the request
+ * @param timeout run onTimeout after this many milliseconds
+ * @param onTimeout the request's startTime and timeout will be passed to this
+ */
+export function createFetchWithNonStoppingTimeout(
+  fetchFunction: typeof fetch,
+  timeout: number,
+  onTimeout: (startTime: number, timeout: number) => void
+): typeof fetch {
+  return async (...[input, init]: Parameters<typeof fetch>) => {
+    const startTime = Date.now();
+    const timer = setTimeout(() => onTimeout(startTime, timeout), timeout);
+
+    try {
+      return await fetchFunction(input, init);
+    } finally {
+      if (timer !== undefined) {
+        clearTimeout(timer);
+      }
+    }
+  };
+}
+
+//
+// HTTPCLIENT UTILS
+//
+
+export function createHttpClient(
+  fetchFunction: typeof fetch,
+  apiKey: string,
+  withCredentials: boolean
+): HttpClient {
+  return async (config: HttpClientConfig) => {
+    let url = config.url;
+    if (config.params) {
+      url = `${url}?${stringify(config.params)}`;
+    }
+
+    const fetchOptions = new Request(url, {
+      method: config.method,
+      body: config.body ? JSON.stringify(config.body) : undefined,
+      headers: config.body
+        ? {
+            'X-API-Key': apiKey,
+            'Content-Type': 'application/json',
+          }
+        : {
+            'X-API-Key': apiKey,
```

Comment:
Did we ever figure out whether there were some calls that didn't need an API key?

Reply:
a handful seem to work ok like global settings endpoint, or usually /Manifest/ but i tihnk that may just be caching & nothing doesn't... not.. not work..? with an api key, except the manifest json & its components